### PR TITLE
Refine file locks for inter process communication

### PR DESCRIPTION
**Pull request by gossion**

### CODE REVIEW

File: src/bosh_azure_cpi/lib/cloud/azure/stemcell_manager2.rb [hunk 1]
```diff
@@ -81,47 +81,27 @@ def get_user_image(stemcell_name, storage_account_type, location)
...
-              stemcell_source_blob_uri = @blob_manager.get_blob_uri(default_storage_account_name, STEMCELL_CONTAINER, "#{stemcell_name}.vhd")
-              @blob_manager.copy_blob(storage_account_name, STEMCELL_CONTAINER, "#{stemcell_name}.vhd", stemcell_source_blob_uri) do
-                mutex.update()
-              end
-            end
-            mutex.unlock
-          else
-            mutex.wait
+        flock("#{CPI_LOCK_COPY_STEMCELL}-#{stemcell_name}-#{storage_account_name}", File::LOCK_EX) do
+          unless has_stemcell?(storage_account_name, stemcell_name)
```
[the remark this answers is not in the document]
Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/stemcell_manager2.rb]: you can "Hide whitespace changes"  in the review page.

[the remark this answers is not in the document]
Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/stemcell_manager2.rb]: cool function.

File: src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb [hunk 2]
```diff
@@ -474,62 +481,41 @@ def get_availability_set_name(resource_pool, env)
...
-            @logger.info("create_availability_set - the availability set `#{availability_set_name}' exists. No need to update.")
-            return availability_set
-          end
-        end
-      end
-
-      mutex = FileMutex.new("#{CPI_LOCK_PREFIX_AVAILABILITY_SET}-create-#{availability_set_name}", @logger)
-      begin
-        if mutex.lock
+        elsif !availability_set[:location].casecmp(availability_set_params[:location]).zero?
```
Comment (andyliuliming) [inline src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb]: is this change related to the issue customer reported?

Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb]: yes, there can be the same error in create_vm.

Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb]: I mean the changes like:
1. @logger.info("The availability set `#{availability_set_name}' doesn't exist. Will create a new one.")
2. casecmp
3. elsif  adjustment.

Comment (andyliuliming) [reply src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb]: I ask this is because basicially the hotfix should contains as minor code change as possible.

Comment (gossion) [reply src/bosh_azure_cpi/lib/cloud/azure/vm_manager.rb]: Got your point, this function is part of #414 (except the logger info) which was already tested, and using a tested function is safer than hijack it.
